src/Polysemy/Internal/TH/Effect.hs

```diff
@@ -75,135 +79,185 @@ makeSem = genFreer True
+  , conArgs  :: [Type]
+  } deriving Show
+
+------------------------------------------------------------------------------
+-- | Creates info about constructor being lifted from parent type's and it's
+-- own info.
+mkCLInfo :: DatatypeInfo -> ConstructorInfo -> Q ConLiftInfo
+-- mkCLInfo dtInfo conInfo = do
+mkCLInfo DatatypeInfo {
+             datatypeName = typeName
+           , datatypeVars
+           , datatypeInstTypes
+           }
+         ConstructorInfo {
+             constructorName = conName
+           , constructorContext
+           , constructorFields
+           }
+  = do
+  let funName           = mapNameBase (mapHead toLower) conName
+      -- We extract equality constraints with variables to unify them
+      -- manually - this makes type errors more readable
+      (eqPairs, conCxt) = first (M.fromList . maybeAddResKind)
+                        $ partitionWith eqPairOrCxt constructorContext
+      -- We constrain kind of result to 'Type' if we get a type variable
+      -- TODO: we probably want to add requirement for KindSignatures into
+      -- documentation - PolyKinds are already mentioned...
+      -- TODO: This makes function signatures ugly - what is the best way to
+      -- restrict occurence of kind signature or completely remove it?
+      -- TODO: take result type by name or position?
+      maybeAddResKind   = maybe id ((:) . (, StarT))
+                        $ tvName' $ tvKind $ last datatypeVars
+      allEffTArgs       = applySubstitution eqPairs datatypeInstTypes
+      conArgs           = applySubstitution eqPairs constructorFields
+
+  (effTArgs, [mArg, resArg]) <- case splitFromEnd 2 allEffTArgs of
+      r@(_, [_, _]) -> return r
+      _             -> fail $ show $ missingEffTArgs typeName
+  mName <-
+    -- May happen when used on data families
+    maybe (fail $ show $ mArgNotVar typeName mArg) return $ tvName' mArg
+
+  return CLInfo {
+    typeName, funName, conName, effTArgs, mName, resArg, conCxt, conArgs
+    }
+  where
+    mArgNotVar name mArg
+      =   text "Monad argument ‘" <> ppr mArg <> text "’ in effect ‘"
+          <> ppr name <> text "’ is not a type variable"
+
+    missingEffTArgs name
+      =   text "Effect ‘" <> ppr name
+          <> text "’ has not enough type arguments"
+      $+$ nest 4
+          (   text "At least monad and result argument are required, e.g.:"
+          $+$ nest 4
+              (   text ""
+              $+$ ppr (DataD [] base args Nothing [] []) <+> text "..."
+              $+$ text ""
+              )
+          )
+      where
+        base = mapNameBase id name
+        args = PlainTV . mkName <$> ["m", "a"]
+
+------------------------------------------------------------------------------
+-- | Generates signature for lifting function and type arguments to apply in
+-- it's body on effect's data constructor.
+genSig :: ConLiftInfo -> Q Dec
+genSig CLInfo {
+    typeName, funName, effTArgs, mName, resArg, conCxt, conArgs
```

Comment:
Nit: I realize it's not particularly widespread around the codebase, but I'd like to start naming local variables with `snake_case`, leaving `camelCase` for top-level things.

Reply:
Ok.